### ISSUE TITLE
Separate game mode code into server and client side code versions.

### ISSUE THREAD

**Issue by IAmNotHanni** [inexorgame/inexor-core#339]
Until now the overall architecture of gamemode header files is the following:
you have one file (like ctf.hpp) which contains the game mode code for both sides: server and client.
**Conditional compiling** is used to generate server and client side code like in this invented example:

```
class CTF
{
    clientinfo get_flag_holder();

    #ifdef SERVMODE
        void kick_flag_holder();
    #else 
        void takeflag();
        void dropflag();
    #endif
};
```

this conditional compiling block represents an **inheritage relation**. This means there should be a class concept for this. The elaborated code block would be the following:

```
class CTF
{
    clientinfo get_flag_holder();
};

class CTF_serverside : public CTF
{
    void kick_flag_holder();
};

class CTF_clientside : public CTF
{
    void takeflag();
    void dropflag();
};
```

Also these classes could be separated into 3 files.
This would be a great replacement for the #define SERVMODE header file includes that can be found everywhere in the code.

**Comment by Fohlen** [comment]
I did thought about this for quiet a while now .. can we do the following:

- implement a `pseudo-gamemode``
- add a `naming` or `identify` property for the gamemode
- let it up to the `tree` to decide the entire settings of the gamemode (e.g: physics, teammode ...)
- when a player joins a server, it's mode tree information is synchronized via the `api`
- that's it. single gamemode, unlimited possibilites.

**Comment by Croydon** [comment]
Before we can do something like that we have to find ways how to prevent cheating.

**Comment by Fohlen** [comment]
Seeking them on the client side in a truly hackable game seems clearly the wrong path to go down..

**Comment by Croydon** [comment]
I didn't say a word of implementing anti cheats methods in the client.
Still, this is going heavily towards scriptable game modes and if we implement that into the client without a plan for anti-cheat, it will be even more easy to cheat for quite some time. Not something to look forward to.

**Comment by Fohlen** [comment]
I do think that is the price we pay for a truly modable game .. and implementing anti cheat is non of public business .. @aschaeffer has been measuring anti cheat systems for server side which do work quiet well. On the other hand the server is as well modable making it easier to write comprehensive anti cheat systems

Am 12.01.2017, 23:23, Croydon <notifications@github.com> schrieb:

I didn't say a word of implementing anti cheats methods in the client.
Still, this is going heavily towards scriptable game modes and if we implement that into the client without a plan for anti-cheat, it will be even more easy to cheat for quite some time. Not something to look forward to.

**Comment by Croydon** [comment]
>  and implementing anti cheat is non of public business

I disagree. The implementation is open so should be it's development.

But we are getting more and more Off-Topic. I just wanted to note, that we should do something to make cheating harder before a stable release, when we currently making it even more easier.

**Comment by aschaeffer** [comment]
Wouldn't we avoid that discussions about cheats and anti cheats? At least until we get to a point where it makes sense.

**Comment by Fohlen** [comment]
What is with the idea of a "mother" game mode, does it sound reasonable?

**Comment by Fohlen** [comment]
By the way, anti-cheat also refers to #167

**Comment by aschaeffer** [comment]
This ticket is only about the refactoring the mega-strange game modes (`ctf.h`, `hold.h`, ``) which mixes server code and client code and makes them confusing, unclear and indivisible.

Non-goals of this ticket:
* Anti-Cheats
* Mother-Gamemode

In fact this work is a prerequisite before we can implement a dynamic system for game modes (aka "mother" game mode).

**Comment by a-teammate** [comment]
Done with #455
